### PR TITLE
Support truncation_mode in SFT

### PR DESCRIPTION
**Pull request by albertvillanova**
Support `truncation_mode` in SFT.

This PR aligns SFT with DPO, as a follow-up to:
- https://github.com/huggingface/trl/pull/5286#issuecomment-4073078242

This PR introduces the `truncation_mode` option in SFT, allowing users to choose whether to keep the start or the end of sequences when truncating. The changes update configuration, documentation, and add validation for vision-language models to ensure safe usage. 

### Changes

**Feature: Truncation Mode Support**
* Added the `truncation_mode` parameter to the `truncate_dataset` function, enabling users to specify whether to keep the start (`"keep_start"`) or end (`"keep_end"`) of sequences when truncating.
* Modified the truncation logic to handle both modes using PyArrow slicing, ensuring the correct part of the sequence is kept based on the mode.

**Configuration and Documentation Updates**
* Introduced the `truncation_mode` field in `SFTConfig`, updated help text and metadata, and clarified how `max_length` interacts with truncation mode.

**Vision-Language Model Safety**
* Added validation in `SFTTrainer` to prevent using `truncation_mode='keep_end'` for vision-language models, since this can drop image tokens and cause mismatches.

**Integration**
* Updated the `tokenize_fn` logic to pass the new `truncation_mode` argument to `truncate_dataset`, ensuring the mode is respected during tokenization.

---

> [!NOTE]
> **Medium Risk**
> Changes SFT preprocessing/truncation behavior by introducing a new user-facing `truncation_mode` and new Arrow-based truncation path (including edge cases like `max_length=0`), which could affect training inputs; adds a VLM guardrail to prevent silent image-token drops.
> 
> **Overview**
> Adds a new `truncation_mode` option for dataset truncation so callers can keep either the start (`keep_start`, default) or end (`keep_end`) of overlong sequences.
> 
> Updates SFT to expose this via `SFTConfig.truncation_mode`, pass it through to `truncate_dataset` during preprocessing, and block `keep_end` for vision-language models to avoid dropping image tokens. Tests are expanded to cover `keep_end` behavior and the `max_length=0` edge case.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit fa7f7f8e8c6b814599d5f66d8f9b728355e4eddc. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: trl/data_utils.py [hunk 1]
```diff
@@ -877,7 +877,12 @@ def pack_dataset(
     return dataset
 
 
-def truncate_dataset(dataset: DatasetType, max_length: int, map_kwargs: dict[str, Any] | None = None) -> DatasetType:
+def truncate_dataset(
+    dataset: DatasetType,
+    max_length: int,
+    map_kwargs: dict[str, Any] | None = None,
+    truncation_mode: str = "keep_start",
```
Comment (qgallouedec) [inline trl/data_utils.py]: turbo nit: can you switch the order of these two args? Having kwargs at the end sounds more natural

Comment (albertvillanova) [reply trl/data_utils.py]: I agree @qgallouedec : I thought the same when implementing, but then I realized that would break backward compatibility when args are passed as positional args (not as keyword args): 
```python
truncate_dataset(dataset, args.max_length, map_kwargs, args.truncation_mode)
```

But maybe we should not care about backward compatibility in this case?

File: trl/data_utils.py [hunk 2]
```diff
@@ -907,14 +914,23 @@ def truncate_dataset(dataset: DatasetType, max_length: int, map_kwargs: dict[str
...
             if pyarrow.types.is_list(column.type) or pyarrow.types.is_large_list(column.type):
-                column = pc.list_slice(column, 0, max_length)
+                if truncation_mode == "keep_start":
+                    column = pc.list_slice(column, 0, max_length)
+                else:  # keep_end
+                    column = (
+                        pa.array([[] for _ in range(len(column))], type=column.type)
+                        if max_length == 0
+                        else pa.array([values[-max_length:] for values in column.to_pylist()], type=column.type)
+                    )
```
[the remark this answers is not in the document]
Comment (albertvillanova) [reply trl/data_utils.py:933]: It doesn't:
```python
_________________________________________________________________________________ TestTruncateExamples.test_with_keep_end _________________________________________________________________________________

self = <tests.test_data_utils.TestTruncateExamples object at 0x7f44fbcfd0a0>

    def test_with_keep_end(self):
        examples = {
            "input_ids": [[1, 2, 3], [4, 5, 6, 7], [8]],
            "attention_mask": [[0, 1, 1], [0, 0, 1, 1], [1]],
        }
        dataset = Dataset.from_dict(examples)
        expected_output = {
            "input_ids": [[2, 3], [6, 7], [8]],
            "attention_mask": [[1, 1], [1, 1], [1]],
        }
>       dataset = truncate_dataset(dataset, max_length=2, truncation_mode="keep_end")
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

tests/test_data_utils.py:1258: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
trl/data_utils.py:939: in truncate_dataset
    dataset = dataset.map(truncate, batched=True, **map_kwargs)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.12/site-packages/datasets/arrow_dataset.py:562: in wrapper
    out: Union["Dataset", "DatasetDict"] = func(self, *args, **kwargs)
                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.12/site-packages/datasets/arrow_dataset.py:3343: in map
    for rank, done, content in Dataset._map_single(**unprocessed_kwargs):
                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.12/site-packages/datasets/arrow_dataset.py:3699: in _map_single
    for i, batch in iter_outputs(shard_iterable):
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.12/site-packages/datasets/arrow_dataset.py:3649: in iter_outputs
    yield i, apply_function(example, i, offset=offset)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.12/site-packages/datasets/arrow_dataset.py:3572: in apply_function
    processed_inputs = function(*fn_args, *additional_args, **fn_kwargs)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
trl/data_utils.py:932: in truncate
    else pc.list_slice(column, -max_length)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.12/site-packages/pyarrow/compute.py:271: in wrapper
    return func.call(args, options, memory_pool)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pyarrow/_compute.pyx:399: in pyarrow._compute.Function.call
    ???
pyarrow/error.pxi:155: in pyarrow.lib.pyarrow_internal_check_status
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   pyarrow.lib.ArrowInvalid: `start`(-2) should be greater than 0 and smaller than `stop`((nullopt))

pyarrow/error.pxi:92: ArrowInvalid
```

File: trl/data_utils.py (hunk 1 above)
Comment (qgallouedec) [reply trl/data_utils.py]: yes, I'd say we don't care, `truncate_dataset` is internal, see motivation in https://github.com/huggingface/trl/issues/5090
